File: backend/app/api/v1/team.py
"""The `/api/v1/team` endpoint: a Manager's Direct Reports — Manager-only (Story 3.2).

Implements: FR-19, api-contracts §4.9 (`GET /team` | Role `Manager` | Scope `reports`),
AD-10/AD-14 (the scope is a SQL predicate over the live reporting edge, resolved from the
DB actor at request time), G3 (a non-Manager is refused `403 ACTION_NOT_PERMITTED` by the
role gate in the dependency, before any row is read), NFR-11 (the list is page-bounded).

--- The one contract inversion: the ADMIN is refused here ---

Nearly every read in this app grants the Admin `ALL`; §4.9 grants `/team` to the Manager
ALONE, so `require_role(authz.ROLE_MANAGER)` refuses an Admin exactly as it refuses an
Employee (AC4). This is contract, not accident: an Admin sees everyone through
`GET /employees`; a team is a fact about a reporting edge, and only a Manager stands on one.

--- Why `TeamMemberResponse` is deliberately smaller than `EmployeeResponse` ---

FR-19 names exactly three facts a Manager is granted sight of: the Employee (Full Name —
erd.md GAP-2 settled that Full Name IS the identification), their Department, and the
active state. `EmployeeResponse` is the ADMIN view and additionally discloses `email`,
`role`, `joining_date` and `manager_id` — none of which any requirement grants a Manager.
Authority is granted per-surface, in the projection, never by convenience (AD-10), so this
response carries `{id, full_name, department, is_active}` and NOTHING else. `id` is the
uuidv7 React key and any later drill-down's handle; `manager_id` is omitted as tautological
(it is the caller).
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends
from pydantic import BaseModel

from app.api.v1.dependencies import Actor, require_role
from app.api.v1.employees import DepartmentBrief
from app.api.v1.pagination import Page, PageParams
from app.services import authorization as authz
from app.services import team as team_service

router = APIRouter()


class TeamMemberResponse(BaseModel):
    """One Direct Report, as FR-19 grants a Manager sight of them — and nothing more."""

    id: uuid.UUID
    full_name: str
    department: DepartmentBrief
    is_active: bool


def _to_response(member: object) -> TeamMemberResponse:
    """Project the service's returned `Employee` into the response, by hand (Trap 5).

    Typed `object` because `api/` may not import the ORM `Employee`; the service guarantees
    `department` is eager-loaded and readable after the session closes. The fields NOT read
    here — `email`, `role`, `joining_date`, `manager_id`, `password_hash` — are the
    disclosure decision (Open Decision #1); the omission is by construction.
    """
    return TeamMemberResponse(
        id=member.id,
        full_name=member.full_name,
        department=DepartmentBrief(
            id=member.department.id, name=member.department.name
        ),
        is_active=member.is_active,
    )


@router.get("/team", tags=["team"])
def list_team(
    params: PageParams = Depends(),
    manager: Actor = Depends(require_role(authz.ROLE_MANAGER)),
) -> Page[TeamMemberResponse]:
    """Return a page of the caller's Direct Reports (AC1, AC2). Manager-only — an Employee
    AND an Admin are `403` (AC4, the §4.9 inversion).

    The role gate runs in the `manager` dependency, before this body — a non-Manager never
    reaches the read (G3). The service applies the REPORTS scope in the SQL; a deactivated
    report is in the list, distinguishable by `is_active` (AC3). The page is bounded by
    `PageParams` (NFR-11); the body carries the `items`/`page`/`page_size`/`total` envelope.
    """
    rows, total = team_service.list_team(params.limit, params.offset, manager)
    return Page[TeamMemberResponse](
        items=[_to_response(row) for row in rows],
        page=params.page,
        page_size=params.page_size,
        total=total,
    )

File: frontend/src/api/team.ts
/**
 * The Manager's team, as a typed hook on `apiFetch`. `GET /api/v1/team` (Story 3.2).
 *
 * Implements: FR-19 (frontend) — a Manager's Direct Reports, each identified by Full Name
 * and Department, deactivated reports present and distinguishable. Manager-ONLY server-side
 * (api-contracts §4.9): an Employee AND an Admin get a `403` — the one read in the app
 * where the Admin is refused — so the panel gates the fetch with `enabled` and the server's
 * role gate stays the real guard. The response is the minimal disclosure the backend chose
 * (Open Decision #1): no email, no role, no joining date reaches this client.
 */
import { useQuery } from '@tanstack/react-query'

import { apiFetch } from './client'
import type { Page } from './departments'
import type { DepartmentBrief } from './me'

/**
 * One Direct Report on the wire — EXACTLY `{id, full_name, department, is_active}`,
 * mirroring the backend `TeamMemberResponse`. `department` reuses the `{id, name}` brief
 * (`me.ts` is its single home, as `employees.py` is the backend's).
 */
export interface TeamMember {
  id: string
  full_name: string
  department: DepartmentBrief
  is_active: boolean
}

/**
 * The base cache key for the team list. Each page caches under `[...TEAM_QUERY_KEY, params]`
 * (TanStack v5 hashes keys structurally), so prefix invalidation would reach every page —
 * though nothing invalidates it yet: this is a read-only surface with no mutation.
 */
export const TEAM_QUERY_KEY = ['team'] as const

/**
 * The Manager's Direct Reports, one page at a time (FR-19). `page`/`page_size` are the only
 * params — the endpoint has no filters — and every value is `encodeURIComponent`-escaped
 * into the query string (the 2.7 review's rule). `options.enabled` gates the fetch (the
 * `useLeaveRequests` idiom): a non-Manager, which renders nothing, never issues a request
 * that is guaranteed to 403.
 */
export function useTeam(
  params: { page?: number; pageSize?: number } = {},
  options?: { enabled?: boolean },
) {
  const pairs: string[] = []
  if (params.page !== undefined) {
    pairs.push(`page=${encodeURIComponent(String(params.page))}`)
  }
  if (params.pageSize !== undefined) {
    pairs.push(`page_size=${encodeURIComponent(String(params.pageSize))}`)
  }
  const path: `/${string}` = pairs.length > 0 ? `/team?${pairs.join('&')}` : '/team'
  return useQuery({
    queryKey: [...TEAM_QUERY_KEY, params],
    queryFn: () => apiFetch<Page<TeamMember>>(path),
    enabled: options?.enabled,
  })
}
